tree/tree/inc/ROOT/TThreadedTree.hxx

```diff
@@ -0,0 +1,109 @@
+#ifndef ROOT_TThreadedTTree
+#define ROOT_TThreadedTTree
+
+#include "ROOT/TThreadedObject.hxx"
+#include "ROOT/TFileMergeClient.hxx"
+#include "ROOT/TFileMergeServer.hxx"
+
+#include "TError.h"
+#include "TTree.h"
+
+#include <memory>
+
+namespace ROOT {
+
+template <>
+class TThreadedObject<TTree> {
+public:
+   using TFileMergeClient = ROOT::Experimental::TFileMergeClient;
+   using TFileMergeServer = ROOT::Experimental::TFileMergeServer;
+
+   TThreadedObject(const char *name, const char *title, int splitlevel = 99)
+      : fName(name), fTitle(title), fSplitLevel(splitlevel)
+   {
+      if (!gDirectory) Fatal("TThreadedObject<TTree>", "Cannot find an open file to which TTree should be associated");
+
+      fFileName = gDirectory->GetName();
+      fMergeServer.reset(new TFileMergeServer(std::string(fFileName.Data())));
+      fMergeClients.reserve(64);
+   }
+
+   std::shared_ptr<TTree> Get() { return GetAtSlot(GetThisSlotNumber()); }
+
+   std::shared_ptr<TTree> GetAtSlot(unsigned i)
+   {
+      if (i >= 64) Fatal("TThreadedObject<TTree>::GetAtSlow()", "Number of clients exceeds maximum number of slots");
```

Comment:
Looks like we could use `TTypeNameExtraction::Get()` introduced by 7f229541 and 1c1bfac6.